spec.bs

```diff
@@ -989,6 +997,43 @@ With the above infrastructure in place, we can actually fire and handle the {{Ap
     :: "{{AppHistoryNavigationType/push}}"
 </div>
 
+<h3 id="navigate-event-traversal-patches">History traversal updates</h3>
+
+<div algorithm="traverse the history by a delta">
+  The <a spec="HTML">traverse the history by a delta</a> algorithm will be totally re-written as part of <a href="https://github.com/whatwg/html/pull/6315">the session history rewrite</a>. Here we reproduce the final version of the algorithm, after both that rewrite and with appropriate app history updates. Recall that per [[#user-initiated-patches]] we have introduced a |userInvolvement| argument, so the arguments are now |delta|, |source browsing context|, and |userInvolvement|.
+
+  1. Let |traversable| be |source browsing context|'s [=browsing context/containing navigable=]'s [=navigable/traversable navigable=].
+  1. Let |initiatorOrigin| be |source browsing context|'s [=active document=]'s [=Document/origin=].
+  1. [=parallel queue/enqueue steps|Enqueue the following steps=] to |traversable|'s [=traversable navigable/session history traversal queue=]:
+    1. Let |allSteps| be the result of [=traversable navigable/getting all history steps=] for |traversable|.
+    1. Let |targetStep| be TODO compute using [=traversable navigable/current session history step=], |allSteps|, and |delta|. Will be failure if there is no step |delta| away from the current one.
+    1. If |targetStep| is failure, then return.
+    1. [=Apply the history step=] |targetStep| to |traversable| with true, <var ignore>step</var>, |initiatorOrigin|, and |userInvolvement|.
+</div>
+
+<div algorithm="apply the history step">
+  Modify the <a spec="HTML">apply the history step</a> algorithm as follows. Change <var ignore>checkForUserCancellation</var> to |fireBeforeunloadAndNavigate|. Add the |userInvolvement| parameter. Then, insert the following step after step 12 (which assembles |toTraverse| and other lists) but before step 13 (which checks if unloading is user-cancelled):
+
+  1. If |fireBeforeunloadAndNavigate| is true and the result of <a>firing traversal `navigate` events</a> given |toTraverse|, <var ignore>step</var>, <var ignore>initiatorToCheck</var>, and |userInvolvement| is false, then return.
```

Comment:
Nit: It seems weird to me when an algorithm returns a boolean like this, but the name of the algorithm doesn't suggest what the return means. That's why I went for "checking if unloading is user-cancelled", but I guess the flip side of that is it isn't clear that the algorithm also fires events.

Reply:
Yeah, I agree it's a bit iffy but I'm modeling it after the existing "fire an event" algorithm.